File: backend/lite/msp.py
# Copyright (c) 2013, 2014, 2015 MIT Probabilistic Computing Project.
#
# This file is part of Venture.
#
# Venture is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# Venture is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Venture.  If not, see <http://www.gnu.org/licenses/>.

from venture.lite.env import VentureEnvironment
from venture.lite.psp import DeterministicPSP, ESRRefOutputPSP
from venture.lite.request import Request,ESR
from venture.lite.sp import SP, VentureSPRecord
from venture.lite.sp import SPType
from venture.lite.sp_help import typed_nr
from venture.lite.sp_registry import registerBuiltinSP
import venture.lite.address as addr
import venture.lite.types as t

class MakeMSPOutputPSP(DeterministicPSP):
  def simulate(self,args):
    sharedOperatorNode = args.operandNodes[0]
    return VentureSPRecord(SP(MSPRequestPSP(sharedOperatorNode),ESRRefOutputPSP()))

  def description(self,name):
    return "%s returns the stochastically memoized version of the input SP." % name

class MSPRequestPSP(DeterministicPSP):
  def __init__(self,sharedOperatorNode):
    self.sharedOperatorNode = sharedOperatorNode

  def simulate(self,args):
    vals = args.operandValues()
    id = str(vals)
    exp = ["memoizedSP"] + [["quote",val] for val in vals]
    env = VentureEnvironment(None,["memoizedSP"],[self.sharedOperatorNode])
    return Request([ESR(id,exp,addr.req_frame(id),env)])

registerBuiltinSP("mem",typed_nr(MakeMSPOutputPSP(),
                                 [SPType([t.AnyType("a")], t.AnyType("b"), variadic=True)],
                                 SPType([t.AnyType("a")], t.AnyType("b"), variadic=True)))

File: backend/new_cxx/src/sps/msp.cxx
// Copyright (c) 2014 MIT Probabilistic Computing Project.
//
// This file is part of Venture.
//
// Venture is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// Venture is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with Venture.  If not, see <http://www.gnu.org/licenses/>.

#include "sps/msp.h"
#include "sprecord.h"
#include "env.h"
#include "utils.h"

VentureValuePtr MakeMSPOutputPSP::simulate(shared_ptr<Args> args, gsl_rng * rng) const
{
  checkArgsLength("mem", args, 1);
  return VentureValuePtr(new VentureSPRecord(new SP(new MSPRequestPSP(args->operandNodes[0]), new ESRRefOutputPSP())));
}

MSPRequestPSP::MSPRequestPSP(Node * sharedOperatorNode) : sharedOperatorNode(sharedOperatorNode) {}

VentureValuePtr quote(const VentureValuePtr& v)
{
  vector<VentureValuePtr> exp;
  exp.push_back(VentureValuePtr(new VentureSymbol("quote")));
  exp.push_back(v);
  return VentureValuePtr(new VentureArray(exp));
}

VentureValuePtr MSPRequestPSP::simulate(shared_ptr<Args> args, gsl_rng * rng) const
{
  vector<VentureValuePtr> exp;
  exp.push_back(shared_ptr<VentureSymbol>(new VentureSymbol("memoizedSP")));
  
  for (size_t i = 0; i < args->operandValues.size(); ++i)
  {
    exp.push_back(quote(args->operandValues[i]));
  }
  
  shared_ptr<VentureEnvironment> empty(new VentureEnvironment());
  empty->addBinding("memoizedSP",sharedOperatorNode);
  
  vector<ESR> esrs;
  esrs.push_back(ESR(VentureValuePtr(new VentureArray(args->operandValues)),VentureValuePtr(new VentureArray(exp)),empty));
  
  return VentureValuePtr(new VentureRequest(esrs, vector<shared_ptr<LSR> >()));
}
